### PR TITLE
feat(ooda): Stages 1–5 — OODA Loop Plugin idea, research, requirements, design, spec

### PR DESCRIPTION
**Pull request by Luis85**
## Summary

- Stage 1 (Idea): `IDEA-OODA-001` — accepted
- Stage 2 (Research): `RESEARCH-OODA-001` — complete (5 parallel agents, 2 passes)
- Stage 3 (Requirements): `PRD-OODA-001` — accepted (32 EARS requirements, 4 OQs resolved)
- Stage 4 (Design): `DESIGN-OODA-001` — accepted (UX flows, UI patterns, full architecture)
- Stage 5 (Specification): `SPEC-OODA-001` — accepted (1,501 lines, 57 test scenarios)

## Spec lineage

| Stage | Artifact |
|---|---|
| Idea | `specs/ooda-loop-plugin/idea.md` |
| Research | `specs/ooda-loop-plugin/research.md` |
| Requirements | `specs/ooda-loop-plugin/requirements.md` |
| Design | `specs/ooda-loop-plugin/design.md` |
| Specification | `specs/ooda-loop-plugin/spec.md` |
| Tasks | *(pending)* |

## Specification highlights (SPEC-OODA-001)

11 SPECDOC-OODA interface items covering all 32 REQs + 7 NFRs:

| Item | Covers |
|---|---|
| SPECDOC-OODA-001 | `/ooda:brief` entry point — 15-step startup sequence, 5 error codes |
| SPECDOC-OODA-002 | First-run wizard — HTTPS/SSH URL parsing, placeholder fallback |
| SPECDOC-OODA-003 | Observe — parallel AgentDefinition dispatch, 30s timeout, on_failure semantics |
| SPECDOC-OODA-004 | Orient — belief decay (−0.2/7d), Pinned Constraints invariant, first-run init |
| SPECDOC-OODA-005 | Decide — tier1_operation/target/params schema, ranking priority order |
| SPECDOC-OODA-006 | Brief rendering — section order, empty-states, filename collision algorithm |
| SPECDOC-OODA-007 | Tier 1 Act — serial exec, 60s undo state machine, GitHub MCP tool↔reversal map |
| SPECDOC-OODA-008 | Feedback + JSONL — atomic `.tmp` rename, non-blocking failure path |
| SPECDOC-OODA-009 | Summariser — last-14 constraint, orient_priority adjustment from feedback |
| SPECDOC-OODA-010 | `ooda-sources.yaml` validation — field-level rules + exact error messages |
| SPECDOC-OODA-011 | `settings.json` fragment — allow/deny evaluation order, bypassPermissions blocked |

**4 state machines** (run lifecycle, summariser trigger, Act undo window, first-run wizard)  
**3 data schemas** with field-level constraints (`ooda-sources.yaml`, `state.md`, `events.jsonl`)  
**21 edge cases** (concurrent runs, malformed JSONL, zero enabled sources, 100% source failure, undo reversal failure…)  
**57 test scenarios** traced to REQ IDs (TEST-OODA-001 through TEST-OODA-057)  
All 7 quality gate items green ✓

## ADRs filed

- ADR-0046 — Plugin packaging as standalone group
- ADR-0047 — Two-file hybrid Orient memory

## Next step

`/spec:tasks` — decompose spec into ~½-day TDD-ordered tasks. Optional first: `/spec:analyze` to cross-check spec ↔ requirements ↔ design consistency.

## Related

- Tracks: #502 (OODA Loop Plugin concept issue)
- Also created: #501 (Goal-Oriented Orchestrator Plugin concept)

https://claude.ai/code/session_01Exinx6yb8wLnKxFth88xNs

### CODE REVIEW

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 1]
```diff
@@ -0,0 +1,1196 @@
...
+| `config` | `ooda-sources.yaml` parsed object | Startup | All phases |
+| `observe_results` | `observe_results` object | Observe | Orient |
+| `orient_digest` | `orient_digest` object | Orient | Decide |
+| `decide_plan` | `decide_plan` object | Decide | Act (selection + execution) |
+| `selected_actions` | array of `decide_plan.actions` entries | Act (selection) | Act (execution) |
+| `actions_taken` | array of execution records | Act (execution) | Finalize (JSONL + brief) |
+| `run_ts` | ISO8601 string | Startup | All phases |
+| `failed_sources` | array of `{source_id, error}` | Observe | Finalize |
+
+No cross-run state except `memory/events.jsonl` and `briefs/`.
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md:773]: Fixed: `memory/state.md` added as persistent cross-run state in SPECDOC-OODA-017 (cross-run state now lists `events.jsonl`, `briefs/`, and `state.md`), read/write steps added to SPECDOC-OODA-005 Orient phase, workspace layout updated in SPECDOC-OODA-018, and new test scenarios TEST-OODA-059/060/061 cover the state.md scenarios.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md:773]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 2]
```diff
@@ -0,0 +1,1196 @@
...
+
+```json
+{
+  "<source_id>": [
+    {
+      "number": 42,
+      "title": "string",
+      "state": "open" | "closed",
+      "labels": ["string"],
+      "assignees": ["string"],
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md:227]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 3]
```diff
@@ -0,0 +1,1196 @@
...
+- **Natural language dashboards or BI visualisations** of brief history.
+- **Hosted or cloud-scheduled invocation.**
+
+---
+
+## 2. Interfaces
+
+### SPECDOC-OODA-001 — `/ooda:brief` skill entry point
+
+- **Kind:** Claude Code skill invocation (`plugins/ooda/SKILL.md` entry point)
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 4]
```diff
@@ -0,0 +1,1196 @@
...
+
+2. For each action in `selected_actions` (in `sequence` order):
+
+   a. Display pre-execution notice (design.md Part B §14 exact copy):
+      ```
+      Executing [<sequence>/<total>]: <tier1_operation> on <item_ref>…
+      ```
+
+   b. Call the GitHub MCP tool corresponding to `tier1_operation`:
+
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md:456]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 5]
```diff
@@ -0,0 +1,1196 @@
...
+  ↓
+[Act: selection]
+  none / empty → [Finalize (no actions)]
+  ≥ 1 selected → [Act: execution]
+  ↓
+[Act: execution]
+  (per action: success/fail/undo loop)
+  ↓
+[Finalize]
+  write brief → write JSONL → delete scratch
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md:832]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 6]
```diff
@@ -0,0 +1,1196 @@
...
+`tier1_operation` valid values: `add_label`, `remove_label`, `post_comment`, `add_reviewer`, `create_draft_issue`, `null`.
+
+Required params per operation:
+- `add_label`: `{owner, repo, issue_number, labels: [string]}`
+- `remove_label`: `{owner, repo, issue_number, name: string}`
+- `post_comment`: `{owner, repo, issue_number|pull_number, body: string}`
+- `add_reviewer`: `{owner, repo, pull_number, reviewers: [string]}`
+- `create_draft_issue`: `{owner, repo, title: string}`
+
+For `add_reviewer`: `{reviewer: "@alice"}`
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 7]
```diff
@@ -0,0 +1,1205 @@
...
+
+---
+
+### SPECDOC-OODA-003 — Observe phase
+
+**Purpose:** Collect raw GitHub data for all configured sources in parallel.
+
+**Inputs:** Validated `ooda-sources.yaml` config.
+
+**Outputs:** `observe_results` map (SPECDOC-OODA-004); scratch file `ooda-runs/<ts>/observe.json`.
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 8]
```diff
@@ -0,0 +1,1205 @@
...
+      ```
+      Wait up to 60 seconds for user response.
+      - `y` or `Y` within 60 seconds → call reversal tool; display `"Action reversed."`; record `{action, undone: true}` for JSONL.
+      - `N`, `n`, Enter, or 60-second timeout → display `"Action finalised."`; record `{action, undone: false}` for JSONL.
+   
+   d. If GitHub MCP call fails → display EC-OODA-005 notice; do not present undo prompt; record `{action, failed: true}` for JSONL.
+   
+   e. If reversal MCP call fails → display EC-OODA-007 notice; record `{action, undone: false, undo_attempted: true}` for JSONL.
+
+3. If the reversal tool for `post_comment` (`mcp__github__delete_issue_comment`) is unavailable → display EC-OODA-008 notice instead of undo prompt.
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md:477]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

File: specs/ooda-loop-plugin/spec.md [hunk 9]
```diff
@@ -0,0 +1,1205 @@
...
+  - REQ-OODA-024
+  - REQ-OODA-025
+  - REQ-OODA-026
+  - REQ-OODA-027
+  - REQ-OODA-028
+  - REQ-OODA-029
+  - REQ-OODA-030
+  - REQ-OODA-031
+  - REQ-OODA-032
+  - REQ-OODA-033
```
[the remark this answers is not in the document]
Comment (Luis85) [reply specs/ooda-loop-plugin/spec.md]: @codex address that feedback

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Luis85** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_
